modules/bg96/cellular_bg96_api.c

```diff
@@ -1901,6 +1914,102 @@ static void _dnsResultCallback( cellularModuleContext_t * pModuleContext,
 
+static CellularError_t udpCheckAndConnect( CellularHandle_t cellularHandle,
+                                           CellularSocketHandle_t socketHandle,
+                                           CellularSocketAccessMode_t dataAccessMode,
+                                           const CellularSocketAddress_t * pRemoteSocketAddress,
+                                           uint32_t timeout )
+{
+    bool needSetRemoteAddress = false;
+    CellularError_t cellularStatus = CELLULAR_SUCCESS;
+    CellularUrcEvent_t urcEvent = CELLULAR_URC_EVENT_OTHER;
+    cellularModuleSocketContext_t * pBg96SocketContext = ( cellularModuleSocketContext_t * ) socketHandle->pModemData;
+
+    PlatformMutex_Lock( &pBg96SocketContext->udpSocketConnectMutex );
+
+    /* Check input. */
+    if( socketHandle == NULL )
+    {
+        LogError( ( "udpCheckAndConnect: Invalid socket address" ) );
+        cellularStatus = CELLULAR_INVALID_HANDLE;
+    }
+    else if( socketHandle->socketProtocol != CELLULAR_SOCKET_PROTOCOL_UDP )
+    {
+        LogError( ( "udpCheckAndConnect, Socket protocol not supported %d",
+                    socketHandle->socketProtocol ) );
+        cellularStatus = CELLULAR_UNSUPPORTED;
+    }
+    else if( dataAccessMode != CELLULAR_ACCESSMODE_BUFFER )
+    {
+        LogError( ( "udpCheckAndConnect, Access mode not supported %d",
+                    dataAccessMode ) );
+        cellularStatus = CELLULAR_UNSUPPORTED;
+    }
+    /* Check if it's necessary to call Cellular_SocketConnect. */
+    else if( ( socketHandle->remoteSocketAddress.port == 0 ) && ( pRemoteSocketAddress == NULL ) )
+    {
+        LogError( ( "udpCheckAndConnect, Remote address is not set correctly." ) );
+        cellularStatus = CELLULAR_BAD_PARAMETER;
+    }
+    else if( socketHandle->remoteSocketAddress.port == 0 )
+    {
+        /* Remote info is not set before. Set pRemoteSocketAddress as remote address. */
+        needSetRemoteAddress = true;
+    }
+    else if( pRemoteSocketAddress == NULL )
+    {
+        /* Remote info is set before. Input remote address is set to NULL, so we can reuse address. */
+        needSetRemoteAddress = false;
+    }
+    else if( memcmp( pRemoteSocketAddress, &socketHandle->remoteSocketAddress, sizeof( CellularSocketAddress_t ) ) != 0 )
+    {
+        /* Remote info is set before. And input remote address is changed. */
+        LogError( ( "Cellular_SocketSendTo, Can't change the remote information in one socket handler" ) );
+        cellularStatus = CELLULAR_UNSUPPORTED;
+    }
+    else
+    {
+        /* Remote info is same as previous setting, reuse it directly. */
+        needSetRemoteAddress = false;
+    }
+
+    /* Create a socket for this socket handler. */
+    if( ( cellularStatus == CELLULAR_SUCCESS ) && needSetRemoteAddress )
+    {
+        cellularStatus = Cellular_SocketConnect( cellularHandle, socketHandle, dataAccessMode, pRemoteSocketAddress );
+    }
+
+    /* Wait for connect result. */
```

Comment:
Suggest to elaborate more about waiting for connect result. UDP socket is not required to connect.